psydac/linalg/tests/test_solvers.py

```diff
@@ -54,6 +64,67 @@ def define_data(n, p, matrix_data, dtype=float):
     xe[s:e + 1] = np.random.random(e + 1 - s)
     return(V, A, xe)
 
+class SquareTorus(Mapping):
+
+    _expressions = {'x': 'x1 * cos(x2)',
+                    'y': 'x1 * sin(x2)',
+                    'z': 'x3'}
+    
+    _ldim        = 3
+    _pdim        = 3
+
+class Annulus(Mapping):
+
+    _expressions = {'x': 'x1 * cos(x2)',
+                    'y': 'x1 * sin(x2)'}
+    
+    _ldim        = 2
+    _pdim        = 2
+
+class SinMapping1D(Mapping):
+
+    _expressions = {'x': 'sin((pi/2)*x1)'}
+    
+    _ldim        = 1
+    _pdim        = 1
+
+def _test_LO_equality_using_rng(A, B):
+    """
+    A simple tool to check with almost certainty that two linear operators are identical, 
+    by applying them repeatedly to random vectors.
+    
+    """
+
+    assert isinstance(A, LinearOperator)
+    assert isinstance(B, LinearOperator)
+    assert A.domain is B.domain
+    assert A.codomain is B.codomain
+
+    rng = np.random.default_rng(42)
+
+    x   = A.domain.zeros()
+    y1  = A.codomain.zeros()
+    y2  = y1.copy()
+
+    n   = 10
+
+    for _ in range(n):
+
+        x *= 0.
+
+        if isinstance(A.domain, BlockVectorSpace):
+            for block in x.blocks:
+                rng.random(size=block._data.shape, dtype="float64", out=block._data)
+        else:
+            rng.random(size=x._data.shape, dtype="float64", out=x._data)
+
+        A.dot(x, out=y1)
+        B.dot(x, out=y2)
+
+        diff = y1 - y2
+        err  = A.codomain.inner(diff, diff)
+        
+        assert err < 1e-15
```

Comment:
Note that `err` is the squared 2-norm of the error, so a tolerance of $10^{-15}$ is not really close to machine precision: $\sqrt{10^{-15}} = 3.16\dotsc\times 10^{-8}$. In any case, we should probably normalize the error by the number of entries in the vector, in order to get an "average error per entry".

I can suggest the following code snippet. Please let me know what you think, and feel free to improve it.
```suggestion
        avg_err_sqr  = A.codomain.inner(diff, diff) / A.codomain.dimension
        TOL = 1e-15
        
        assert avg_err_sqr < TOL**2
```

Reply:
I wanted to avoid taking the square root because it is much more expensive than raising to the squared power, but I forgot to do the same with `A.codomain.dimension`... Your code is equivalent and hence also wrong, I am afraid 😅